clientconn_test.go

```diff
@@ -820,3 +820,126 @@ func TestBackoffCancel(t *testing.T) {
 	cc.Close()
 	// Should not leak. May need -count 5000 to exercise.
 }
+
+// CONNECTING -> READY -> TRANSIENT FAILURE -> CONNECTING -> TRANSIENT FAILURE -> CONNECTING -> TRANSIENT FAILURE
+func TestDialCloseStateTransition(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	server, err := net.Listen("tcp", "localhost:0")
```

Comment:
Nit: this is a listener not a server, so s/server/lis/.  (It's helpful to follow a convention things like this.)